File: utils/transitions.py
import cv2
import numpy as np
from utils.base import BaseTransition, EasingType
from typing import Union, Tuple, Callable

class SlideTransition(BaseTransition):
    def __init__(self, duration: float = 1.0, easing: EasingType = "ease_in_out", direction: str = "left"):
        super().__init__(duration, easing)
        self.direction = direction

    def apply(self, frame1: np.ndarray, frame2: np.ndarray, progress: float) -> np.ndarray:
        h, w = frame1.shape[:2]
        out_frame = np.zeros_like(frame1)

        offset_x, offset_y = 0, 0
        if self.direction == "left":
            offset_x = int(w * progress)
        elif self.direction == "right":
            offset_x = -int(w * progress)
        elif self.direction == "up":
            offset_y = int(h * progress)
        elif self.direction == "down":
            offset_y = -int(h * progress)

        # Draw frame1 shifted out and frame2 shifted in
        if self.direction == "left":
            out_frame[:, :max(0, w - offset_x)] = frame1[:, offset_x:w]
            out_frame[:, max(0, w - offset_x):] = frame2[:, :offset_x]
        elif self.direction == "right":
            # Right means frame1 moves right, frame2 comes from left
            out_frame[:, -offset_x:] = frame1[:, :w + offset_x]
            out_frame[:, :-offset_x] = frame2[:, w + offset_x:]
        elif self.direction == "up":
            out_frame[:max(0, h - offset_y), :] = frame1[offset_y:h, :]
            out_frame[max(0, h - offset_y):, :] = frame2[:offset_y, :]
        elif self.direction == "down":
            out_frame[-offset_y:, :] = frame1[:h + offset_y, :]
            out_frame[:-offset_y, :] = frame2[h + offset_y:, :]

        return out_frame

class ZoomTransition(BaseTransition):
    def __init__(self, duration: float = 1.0, easing: EasingType = "ease_in_out", mode: str = "in"):
        super().__init__(duration, easing)
        self.mode = mode # "in", "out", "inout", "outin"

    def apply(self, frame1: np.ndarray, frame2: np.ndarray, progress: float) -> np.ndarray:
        h, w = frame1.shape[:2]
        
        # Calculate scale factor based on mode
        scale1 = 1.0
        scale2 = 1.0
        alpha = progress

        if self.mode == "in":
            scale1 = 1.0 + progress
            scale2 = 2.0 - progress
        elif self.mode == "out":
            scale1 = 1.0 - (progress * 0.5)
            scale2 = 0.5 + (progress * 0.5)
        elif self.mode == "inout":
            scale1 = 1.0 + progress
            scale2 = 0.5 + (progress * 0.5)
        elif self.mode == "outin":
            scale1 = 1.0 - (progress * 0.5)
            scale2 = 2.0 - progress

        def scale_and_crop(frame, scale):
            nh, nw = int(h * scale), int(w * scale)
            if nh <= 0 or nw <= 0:
                return np.zeros_like(frame)
            resized = cv2.resize(frame, (nw, nh))
            
            # Crop to original size
            if scale > 1.0:
                y1 = (nh - h) // 2
                x1 = (nw - w) // 2
                return resized[y1:y1+h, x1:x1+w]
            # Pad to original size
            else:
                out = np.zeros_like(frame)
                y1 = (h - nh) // 2
                x1 = (w - nw) // 2
                out[y1:y1+nh, x1:x1+nw] = resized
                return out

        f1_zoomed = scale_and_crop(frame1, scale1)
        f2_zoomed = scale_and_crop(frame2, scale2)

        # Crossfade between zoomed frames
        blended = cv2.addWeighted(f1_zoomed, 1.0 - alpha, f2_zoomed, alpha, 0)
        return blended


class GridWipeTransition(BaseTransition):
    """Reveals the next clip in a grid-like block pattern.

    The frame is divided into *cols* × *rows* blocks. Each block
    flips from frame1 to frame2 in a staggered left→right / top→bottom order.

    :param cols: Number of horizontal blocks (default 6).
    :param rows: Number of vertical blocks (default 4).
    :param stagger: Stagger direction — ``"row"`` (left→right per row, default)
                    or ``"col"`` (top→bottom per column).
    :param easing: Easing spec (applied per-block).
    """
    def __init__(
        self,
        duration: float = 1.0,
        easing: EasingType = "ease_in_out",
        cols: int = 6,
        rows: int = 4,
        stagger: str = "row",
    ):
        super().__init__(duration, easing)
        self.cols = max(1, cols)
        self.rows = max(1, rows)
        self.stagger = stagger

    def apply(self, frame1: np.ndarray, frame2: np.ndarray, progress: float) -> np.ndarray:
        h, w = frame1.shape[:2]
        out = frame1.copy()

        bh = h // self.rows
        bw = w // self.cols

        for r in range(self.rows):
            for c in range(self.cols):
                # Stagger: each block has its own local progress
                if self.stagger == "col":
                    idx = c * self.rows + r
                else:
                    idx = r * self.cols + c
                total = self.cols * self.rows
                local_p = (progress * total - idx) / (total - idx)
                local_p = max(0.0, min(1.0, local_p))

                y1 = r * bh
                y2 = (r + 1) * bh if r < self.rows - 1 else h
                x1 = c * bw
                x2 = (c + 1) * bw if c < self.cols - 1 else w

                if local_p >= 1.0:
                    out[y1:y2, x1:x2] = frame2[y1:y2, x1:x2]
                elif local_p > 0.0:
                    a = local_p
                    sub1 = frame1[y1:y2, x1:x2].astype(np.float32)
                    sub2 = frame2[y1:y2, x1:x2].astype(np.float32)
                    blended = cv2.addWeighted(sub1, 1.0 - a, sub2, a, 0)
                    out[y1:y2, x1:x2] = blended.astype(np.uint8)

        return out


class FlashTransition(BaseTransition):
    """Flash to a solid color and back, revealing the next clip.

    Goes frame1 → color → frame2.  Good for beat cuts and energetic edits.

    :param color: BGR tuple for the flash (default (255, 255, 255) = white).
    :param flash_point: Fraction of transition duration at the flash peak
                        (default 0.35 — flash occurs ~1/3 through).
    :param easing: Easing spec.
    """
    def __init__(
        self,
        duration: float = 1.0,
        easing: EasingType = "ease_in_out",
        color: tuple = (255, 255, 255),
        flash_point: float = 0.35,
    ):
        super().__init__(duration, easing)
        self.color = np.array(color, dtype=np.uint8).reshape(1, 1, 3)
        self.flash_point = max(0.05, min(0.95, flash_point))

    def apply(self, frame1: np.ndarray, frame2: np.ndarray, progress: float) -> np.ndarray:
        h, w = frame1.shape[:2]
        fp = self.flash_point

        if progress < fp:
            # Fade frame1 → flash color
            a = progress / fp
            flash = np.full((h, w, 3), self.color, dtype=np.uint8)
            return cv2.addWeighted(frame1, 1.0 - a, flash, a, 0)
        elif progress < 2.0 * fp:
            # Fade flash color → frame2
            a = (progress - fp) / fp
            flash = np.full((h, w, 3), self.color, dtype=np.uint8)
            return cv2.addWeighted(flash, 1.0 - a, frame2, a, 0)
        else:
            return frame2


class RadialWipeTransition(BaseTransition):
    """Radial wipe — a growing circle reveals the next clip from center.

    :param origin: Normalised (x, y) centre of the wipe (default (0.5, 0.5)).
    :param easing: Easing spec.
    """
    def __init__(
        self,
        duration: float = 1.0,
        easing: EasingType = "ease_in_out",
        origin: tuple = (0.5, 0.5),
    ):
        super().__init__(duration, easing)
        self.origin = origin

    def apply(self, frame1: np.ndarray, frame2: np.ndarray, progress: float) -> np.ndarray:
        h, w = frame1.shape[:2]
        cx = int(self.origin[0] * w)
        cy = int(self.origin[1] * h)
        max_radius = np.sqrt(max(cx, w - cx) ** 2 + max(cy, h - cy) ** 2)

        Y, X = np.ogrid[:h, :w]
        dist = np.sqrt((X - cx) ** 2 + (Y - cy) ** 2)
        radius = progress * max_radius

        mask = (dist <= radius).astype(np.float32)

        out = frame1.astype(np.float32) * (1.0 - mask[:, :, np.newaxis])
        out += frame2.astype(np.float32) * mask[:, :, np.newaxis]
        return np.clip(out, 0, 255).astype(np.uint8)


class ZoomInTransition(BaseTransition):
    """Aggressive zoom-in transition — zooms way into frame1, blurs at peak,
    then reveals frame2 underneath.

    Creates a classic "zoom through" feel used in modern video editing.

    :param max_zoom: Peak zoom factor for the outgoing frame (default 5.0).
    :param blur_peak: Max Gaussian blur sigma at midpoint; 0 disables blur (default 3.0).
    :param easing: Easing spec.
    """
    def __init__(
        self,
        duration: float = 0.3,
        easing: EasingType = (0.45, 0, 0.55, 1),
        max_zoom: float = 5.0,
        blur_peak: float = 3.0,
    ):
        super().__init__(duration, easing)
        self.max_zoom = max(max_zoom, 1.01)
        self.blur_peak = blur_peak

    def apply(self, frame1: np.ndarray, frame2: np.ndarray, progress: float) -> np.ndarray:
        h, w = frame1.shape[:2]

        # Zoom: 1× → max_zoom
        scale = 1.0 + (self.max_zoom - 1.0) * progress

        # Alpha: steep curve so frame1 stays dominant most of the time
        alpha = progress ** 3.5

        # Scale + crop frame1
        nh, nw = int(h * scale), int(w * scale)
        resized = cv2.resize(frame1, (nw, nh), interpolation=cv2.INTER_LINEAR)
        y1 = (nh - h) // 2
        x1 = (nw - w) // 2
        f1 = resized[y1:y1 + h, x1:x1 + w]

        # Blur ramp: 0 → peak → 0 (sinusoidal)
        if self.blur_peak > 0.0:
            blur_sigma = self.blur_peak * np.sin(np.pi * progress)
            if blur_sigma > 0.5:
                ksize = int(blur_sigma * 6 + 1) | 1
                f1 = cv2.GaussianBlur(f1, (ksize, ksize), blur_sigma)

        # Crossfade
        return cv2.addWeighted(f1, 1.0 - alpha, frame2, alpha, 0)

File: rust/crates/core/src/transitions.rs
use crate::effect::{EasingHolder, Transition};
use crate::easing::Easing;
use crate::frame::*;
use animato::Interpolate;

pub struct SlideTransition {
    direction: String,
    easing: Easing,
}
impl SlideTransition {
    pub fn new(direction: &str, easing: Easing) -> Self {
        SlideTransition { direction: direction.to_string(), easing }
    }
}
impl EasingHolder for SlideTransition {
    fn easing(&self) -> &Easing {
        &self.easing
    }
}
impl Transition for SlideTransition {
    fn apply(&self, frame1: &Frame, frame2: &Frame, progress: f32) -> Frame {
        let (h, w) = (frame1.height(), frame1.width());
        let mut out = Frame::new(w, h);
        let ox = (w as f32 * progress) as i32;
        let oy = (h as f32 * progress) as i32;
        match self.direction.as_str() {
            "left" => {
                if ox < w as i32 {
                    let sx = ox.max(0) as u32;
                    let sw = (w - sx).min(w);
                    blit_sub(frame1, sx, 0, sw, h, &mut out, 0, 0);
                }
                if ox > 0 {
                    let sw = ox as u32;
                    blit_sub(frame2, 0, 0, sw, h, &mut out, (w as i32 - ox) as u32, 0);
                }
            }
            "right" => {
                let neg = -ox;
                if neg <= 0 {
                    let sw = (w as i32 + neg).max(0) as u32;
                    blit_sub(frame1, 0, 0, sw, h, &mut out, 0, 0);
                }
                if neg <= 0 {
                    let sx = (w as i32 + neg).max(0) as u32;
                    let sw = (-neg).max(0) as u32;
                    blit_sub(frame2, sx, 0, sw, h, &mut out, sx, 0);
                }
            }
            "up" => {
                if oy < h as i32 {
                    let sy = oy.max(0) as u32;
                    let sh = (h - sy).min(h);
                    blit_sub(frame1, 0, sy, w, sh, &mut out, 0, 0);
                }
                if oy > 0 {
                    let sh = oy as u32;
                    blit_sub(frame2, 0, 0, w, sh, &mut out, 0, (h as i32 - oy) as u32);
                }
            }
            "down" => {
                let neg = -oy;
                if neg <= 0 {
                    let sh = (h as i32 + neg).max(0) as u32;
                    blit_sub(frame1, 0, 0, w, sh, &mut out, 0, 0);
                }
                if neg <= 0 {
                    let sy = (h as i32 + neg).max(0) as u32;
                    let sh = (-neg).max(0) as u32;
                    blit_sub(frame2, 0, sy, w, sh, &mut out, 0, sy);
                }
            }
            _ => out = frame2.clone(),
        }
        out
    }
}

pub struct ZoomTransition {
    mode: String,
    easing: Easing,
}
impl ZoomTransition {
    pub fn new(mode: &str, easing: Easing) -> Self {
        ZoomTransition { mode: mode.to_string(), easing }
    }
}
impl EasingHolder for ZoomTransition {
    fn easing(&self) -> &Easing {
        &self.easing
    }
}
impl Transition for ZoomTransition {
    fn apply(&self, frame1: &Frame, frame2: &Frame, progress: f32) -> Frame {
        let (h, w) = (frame1.height(), frame1.width());
        let (s1, s2) = match self.mode.as_str() {
            "in" => (1.0 + progress, 2.0 - progress),
            "out" => (1.0 - progress * 0.5, 0.5 + progress * 0.5),
            "inout" => (1.0 + progress, 0.5 + progress * 0.5),
            "outin" => (1.0 - progress * 0.5, 2.0 - progress),
            _ => (1.0, 1.0),
        };
        let f1 = scale_about_crop(frame1, s1);
        let f2 = scale_about_crop(frame2, s2);
        add_weighted(&f1, 1.0 - progress, &f2, progress)
    }
}

fn scale_about_crop(frame: &Frame, scale: f32) -> Frame {
    let (h, w) = (frame.height(), frame.width());
    let nw = (w as f32 * scale).max(1.0) as u32;
    let nh = (h as f32 * scale).max(1.0) as u32;
    let r = resize_frame(frame, nw, nh, "fill");
    if scale >= 1.0 {
        let x1 = ((nw - w) / 2) as u32;
        let y1 = ((nh - h) / 2) as u32;
        let mut out = Frame::new(w, h);
        blit_sub(&r, x1, y1, w, h, &mut out, 0, 0);
        out
    } else {
        let mut out = Frame::new(w, h);
        let ox = ((w - nw) / 2) as u32;
        let oy = ((h - nh) / 2) as u32;
        blit_sub(&r, 0, 0, nw, nh, &mut out, ox, oy);
        out
    }
}

pub struct GridWipeTransition {
    cols: u32,
    rows: u32,
    stagger: String,
    easing: Easing,
}
impl GridWipeTransition {
    pub fn new(cols: u32, rows: u32, stagger: &str, easing: Easing) -> Self {
        GridWipeTransition { cols, rows, stagger: stagger.to_string(), easing }
    }
}
impl EasingHolder for GridWipeTransition {
    fn easing(&self) -> &Easing {
        &self.easing
    }
}
impl Transition for GridWipeTransition {
    fn apply(&self, frame1: &Frame, frame2: &Frame, progress: f32) -> Frame {
        let (h, w) = (frame1.height(), frame1.width());
        let mut out = Frame::new(w, h);
        let bh = h / self.rows;
        let bw = w / self.cols;
        let total = (self.cols * self.rows) as f32;
        for r in 0..self.rows {
            for c in 0..self.cols {
                let idx = if self.stagger == "col" {
                    c * self.rows + r
                } else {
                    r * self.cols + c
                } as f32;
                let local_p = ((progress * total - idx) / (total - idx)).clamp(0.0, 1.0);
                let y1 = r * bh;
                let y2 = if r < self.rows - 1 { (r + 1) * bh } else { h };
                let x1 = c * bw;
                let x2 = if c < self.cols - 1 { (c + 1) * bw } else { w };
                let cw = x2 - x1;
                let ch = y2 - y1;
                if local_p >= 1.0 {
                    blit_sub(frame2, x1, y1, cw, ch, &mut out, x1, y1);
                } else if local_p > 0.0 {
                    // Blend sub-regions directly without extracting intermediate frames
                    let src1 = frame1.as_raw();
                    let src2 = frame2.as_raw();
                    let dst = out.raw_mut();
                    let fw = w as usize;
                    for yy in y1..y2 {
                        for xx in x1..x2 {
                            let i = (yy as usize * fw + xx as usize) * 3;
                            let be = 1.0 - local_p;
                            dst[i]     = (src1[i] as f32 * be + src2[i] as f32 * local_p).clamp(0.0, 255.0) as u8;
                            dst[i + 1] = (src1[i + 1] as f32 * be + src2[i + 1] as f32 * local_p).clamp(0.0, 255.0) as u8;
                            dst[i + 2] = (src1[i + 2] as f32 * be + src2[i + 2] as f32 * local_p).clamp(0.0, 255.0) as u8;
                        }
                    }
                } else {
                    // local_p == 0: copy from frame1
                    blit_sub(frame1, x1, y1, cw, ch, &mut out, x1, y1);
                }
            }
        }
        out
    }
}

pub struct FlashTransition {
    color: [u8; 3],
    flash_point: f32,
    easing: Easing,
}
impl FlashTransition {
    pub fn new(color: [u8; 3], flash_point: f32, easing: Easing) -> Self {
        let fp = flash_point.clamp(0.05, 0.95);
        FlashTransition { color, flash_point: fp, easing }
    }
}
impl EasingHolder for FlashTransition {
    fn easing(&self) -> &Easing {
        &self.easing
    }
}
impl Transition for FlashTransition {
    fn apply(&self, frame1: &Frame, frame2: &Frame, progress: f32) -> Frame {
        let fp = self.flash_point;
        let [cr, cg, cb] = self.color;
        // Blend directly against the constant flash colour without allocating
        // a solid-colour Frame.
        let blend_to_color = |f: &Frame, a: f32| -> Frame {
            let src = f.as_raw().as_slice();
            let be = 1.0 - a;
            let mut raw = vec![0u8; src.len()];
            for (i, o) in raw.iter_mut().enumerate() {
                let channel_const = match i % 3 { 0 => cb, 1 => cg, _ => cr } as f32;
                *o = (src[i] as f32 * be + channel_const * a).clamp(0.0, 255.0) as u8;
            }
            Frame::from_raw(f.width(), f.height(), raw).expect("flash blend")
        };

        if progress < fp {
            let a = progress / fp;
            blend_to_color(frame1, a)
        } else if progress < 2.0 * fp {
            let a = 1.0 - (progress - fp) / fp;
            blend_to_color(frame2, a)
        } else {
            frame2.clone()
        }
    }
}

pub struct RadialWipeTransition {
    origin: (f32, f32),
    easing: Easing,
}
impl RadialWipeTransition {
    pub fn new(origin: (f32, f32), easing: Easing) -> Self {
        RadialWipeTransition { origin, easing }
    }
}
impl EasingHolder for RadialWipeTransition {
    fn easing(&self) -> &Easing {
        &self.easing
    }
}
impl Transition for RadialWipeTransition {
    fn apply(&self, frame1: &Frame, frame2: &Frame, progress: f32) -> Frame {
        let (h, w) = (frame1.height(), frame1.width());
        let cx = (self.origin.0 * w as f32) as i32;
        let cy = (self.origin.1 * h as f32) as i32;
        let max_r = (((cx as f32 - 0.0).max((w as i32 - cx) as f32)).powi(2)
            + ((cy as f32).max((h as i32 - cy) as f32)).powi(2))
            .sqrt();
        let radius = progress * max_r;
        let mut out = Frame::new(w, h);
        let od = out.raw_mut();
        let a = frame1.as_raw().as_slice();
        let b = frame2.as_raw().as_slice();
        for y in 0..h {
            for x in 0..w {
                let dist = (((x as i32 - cx) as f32).powi(2) + ((y as i32 - cy) as f32).powi(2)).sqrt();
                let sel = if dist <= radius { b } else { a };
                let o = ((y * w + x) * 3) as usize;
                od[o..o + 3].copy_from_slice(&sel[o..o + 3]);
            }
        }
        out
    }
}

pub struct ZoomInTransition {
    max_zoom: f32,
    blur_peak: f32,
    easing: Easing,
}
impl ZoomInTransition {
    pub fn new(max_zoom: f32, blur_peak: f32, easing: Easing) -> Self {
        ZoomInTransition { max_zoom: max_zoom.max(1.01), blur_peak, easing }
    }
}
impl EasingHolder for ZoomInTransition {
    fn easing(&self) -> &Easing {
        &self.easing
    }
}
impl Transition for ZoomInTransition {
    fn apply(&self, frame1: &Frame, frame2: &Frame, progress: f32) -> Frame {
        let (h, w) = (frame1.height(), frame1.width());
        let scale = 1.0_f32.lerp(&self.max_zoom, progress);
        let alpha = progress.powf(3.5);
        let nh = (h as f32 * scale).max(1.0) as u32;
        let nw = (w as f32 * scale).max(1.0) as u32;
        let r = resize_frame(frame1, nw, nh, "fill");
        let x1 = ((nh - h) / 2) as u32;
        let y1 = ((nw - w) / 2) as u32;
        let mut f1 = {
            let mut out = Frame::new(w, h);
            blit_sub(&r, y1, x1, w, h, &mut out, 0, 0);
            out
        };
        if self.blur_peak > 0.0 {
            let sigma = self.blur_peak * (std::f32::consts::PI * progress).sin();
            if sigma > 0.5 {
                f1 = gaussian_blur(&f1, sigma);
            }
        }
        add_weighted(&f1, 1.0 - alpha, frame2, alpha)
    }
}
